File: src/new_body/patch_panel.py
"""Centralized mini-patch panel marshaling the surrogate's sensory nodes.

Mirrors section 3 of the EDD: a grounded, high-density panel in the lower
structural base frame routing all subsystem nodes over 40GBASE-T.

This module is the primary extension point for R&D contributors. New sensor
suites can be registered without editing core code (see ``register_subsystem``
and ``PatchPanelBuilder``), so research teams can bolt on experimental nodes.
"""

from collections.abc import Callable, Iterable
from dataclasses import dataclass, field

from .cat8 import Cat8Link

LinkFactory = Callable[[int], Cat8Link]

MAX_PORTS = 48  # research rigs may exceed the canonical 12-port base chassis


@dataclass(frozen=True)
class SubsystemSpec:
    """Declarative description of a subsystem node for the patch panel."""

    name: str
    interface_type: str
    protocol: str
    poe_class: str | None
    port_range: tuple[int, int]  # 1-indexed, inclusive

    def __post_init__(self) -> None:
        start, end = self.port_range
        if not (1 <= start <= end <= MAX_PORTS):
            raise ValueError(
                f"{self.name}: invalid port range {self.port_range} "
                f"(must be 1..{MAX_PORTS}, start <= end)"
            )

    @property
    def port_ids(self) -> list[int]:
        start, end = self.port_range
        return list(range(start, end + 1))


_REGISTRY: dict[str, SubsystemSpec] = {}


def register_subsystem(spec: SubsystemSpec) -> None:
    """Register (or overwrite) a subsystem spec in the global registry.

    R&D teams call this at import time to add experimental nodes. Port ranges
    must not overlap with an already-registered, differently-named spec.
    """
    for existing in _REGISTRY.values():
        if existing.name == spec.name:
            continue
        a_start, a_end = existing.port_range
        b_start, b_end = spec.port_range
        if max(a_start, b_start) <= min(a_end, b_end):
            raise ValueError(
                f"Port range {spec.port_range} for '{spec.name}' overlaps "
                f"'{existing.name}' range {existing.port_range}"
            )
    _REGISTRY[spec.name] = spec


def registered_subsystems() -> list[SubsystemSpec]:
    return list(_REGISTRY.values())


def clear_registry() -> None:
    """Reset the registry (primarily for tests)."""
    _REGISTRY.clear()


@dataclass
class PatchPort:
    """A single keystone-terminated port on the mini-patch panel."""

    port_id: int
    subsystem: str
    interface_type: str
    protocol: str
    poe_class: str | None
    link: Cat8Link | None = None

    def __str__(self) -> str:
        poe = self.poe_class or "Unpowered"
        return (
            f"Port {self.port_id:02d} | {self.subsystem} | "
            f"{self.interface_type} | {self.protocol} | {poe}"
        )


# --- Canonical 12-port base chassis (section 3 of the EDD) -----------------
_BASE_SPECS: list[SubsystemSpec] = [
    SubsystemSpec(
        "Head & Sensory Node",
        "High-Definition Face-Tracking Cameras & Spatial Microphones",
        "40GBASE-T",
        "PoE++ Type 4 (<=30W)",
        (1, 2),
    ),
    SubsystemSpec(
        "Upper Torso & Kinetic",
        "Localized Haptic Feedback Arrays & Expression Servo Controllers",
        "40GBASE-T",
        "PoE++ Type 4 (Up to 90W)",
        (3, 6),
    ),
    SubsystemSpec(
        "Lower Base & Rig",
        "Placing Rig Alignment Encoders & Frame Security Interlocks",
        "40GBASE-T",
        "PoE++ Type 4 (<=15W)",
        (7, 8),
    ),
    SubsystemSpec(
        "Environmental Matrix",
        "Liquid-Cooling Thermal Telemetry & Lung-Style Intake Fans",
        "40GBASE-T",
        "PoE++ Type 4 (Up to 90W)",
        (9, 10),
    ),
    SubsystemSpec(
        "External Umbilical",
        "High-Speed Remote Uplink to Dedicated Workstation Hub",
        "40GBASE-T",
        "Unpowered (Data Only)",
        (11, 12),
    ),
]


def seed_base_registry() -> None:
    """(Re)seed the canonical 12-port base chassis into the registry."""
    for spec in _BASE_SPECS:
        _REGISTRY.setdefault(spec.name, spec)


def base_specs() -> list[SubsystemSpec]:
    """Return a fresh list of the canonical 12-port base chassis specs."""
    return list(_BASE_SPECS)


seed_base_registry()


class PatchPanelBuilder:
    """Programmatic builder for custom research rig layouts.

    Use this when an R&D team needs a bespoke panel instead of the registry
    defaults (e.g. simulated harnesses with non-contiguous port maps).
    """

    def __init__(self) -> None:
        self._specs: list[SubsystemSpec] = []

    def add(self, spec: SubsystemSpec) -> "PatchPanelBuilder":
        self._specs.append(spec)
        return self

    def build(self, link_factory: LinkFactory) -> "PatchPanel":
        return PatchPanel.from_specs(self._specs, link_factory)


@dataclass
class PatchPanel:
    """Grounded mini-patch panel housed in the lower base frame."""

    ports: list[PatchPort] = field(default_factory=list)

    @classmethod
    def from_specs(
        cls, specs: Iterable[SubsystemSpec], link_factory: LinkFactory
    ) -> "PatchPanel":
        ports: list[PatchPort] = []
        for spec in specs:
            for pid in spec.port_ids:
                ports.append(
                    PatchPort(
                        port_id=pid,
                        subsystem=spec.name,
                        interface_type=spec.interface_type,
                        protocol=spec.protocol,
                        poe_class=spec.poe_class,
                        link=link_factory(pid),
                    )
                )
        ports.sort(key=lambda p: p.port_id)
        return cls(ports=ports)

    @classmethod
    def default_layout(cls, link_factory: LinkFactory) -> "PatchPanel":
        """Build the canonical 12-port base chassis from the registry."""
        return cls.from_specs(registered_subsystems(), link_factory)

    def port(self, port_id: int) -> PatchPort:
        for p in self.ports:
            if p.port_id == port_id:
                return p
        raise KeyError(f"No port {port_id} on the patch panel")

    def subsystems(self) -> list[str]:
        seen: list[str] = []
        for p in self.ports:
            if p.subsystem not in seen:
                seen.append(p.subsystem)
        return seen

    def validate(self) -> list[str]:
        issues: list[str] = []
        for p in self.ports:
            if p.link is not None:
                issues.extend(p.link.validate())
        return issues

    def report(self) -> str:
        return "\n".join(f"{p}" for p in self.ports)

File: rust/core/src/patch_panel.rs
//! Centralized mini-patch panel marshaling the surrogate's sensory nodes.
//!
//! Mirrors section 3 of the EDD: a grounded, high-density panel in the lower
//! structural base frame routing all subsystem nodes over 40GBASE-T.
//!
//! This module is the primary extension point for R&D contributors: new sensor
//! suites register via [`register_subsystem`] (or a [`PatchPanelBuilder`])
//! without editing core code.

use std::sync::{Mutex, OnceLock};

use serde::Serialize;

use crate::cat8::Cat8Link;

pub const MAX_PORTS: u32 = 48; // research rigs may exceed the 12-port base

#[derive(Debug, Clone, Serialize)]
pub struct SubsystemSpec {
    pub name: String,
    pub interface_type: String,
    pub protocol: String,
    pub poe_class: Option<String>,
    pub port_range: (u32, u32),
}

impl SubsystemSpec {
    pub fn new(
        name: &str,
        interface_type: &str,
        protocol: &str,
        poe_class: Option<&str>,
        start: u32,
        end: u32,
    ) -> Result<Self, String> {
        if !(1 <= start && start <= end && end <= MAX_PORTS) {
            return Err(format!(
                "{name}: invalid port range ({start},{end}) must be 1..{MAX_PORTS}, start<=end"
            ));
        }
        Ok(Self {
            name: name.to_string(),
            interface_type: interface_type.to_string(),
            protocol: protocol.to_string(),
            poe_class: poe_class.map(str::to_string),
            port_range: (start, end),
        })
    }

    pub fn port_ids(&self) -> Vec<u32> {
        (self.port_range.0..=self.port_range.1).collect()
    }
}

static REGISTRY: OnceLock<Mutex<Vec<SubsystemSpec>>> = OnceLock::new();

fn registry() -> &'static Mutex<Vec<SubsystemSpec>> {
    REGISTRY.get_or_init(|| Mutex::new(Vec::new()))
}

/// Canonical 12-port base chassis (section 3 of the EDD).
pub fn base_specs() -> Vec<SubsystemSpec> {
    vec![
        SubsystemSpec::new(
            "Head & Sensory Node",
            "High-Definition Face-Tracking Cameras & Spatial Microphones",
            "40GBASE-T",
            Some("PoE++ Type 4 (<=30W)"),
            1,
            2,
        )
        .unwrap(),
        SubsystemSpec::new(
            "Upper Torso & Kinetic",
            "Localized Haptic Feedback Arrays & Expression Servo Controllers",
            "40GBASE-T",
            Some("PoE++ Type 4 (Up to 90W)"),
            3,
            6,
        )
        .unwrap(),
        SubsystemSpec::new(
            "Lower Base & Rig",
            "Placing Rig Alignment Encoders & Frame Security Interlocks",
            "40GBASE-T",
            Some("PoE++ Type 4 (<=15W)"),
            7,
            8,
        )
        .unwrap(),
        SubsystemSpec::new(
            "Environmental Matrix",
            "Liquid-Cooling Thermal Telemetry & Lung-Style Intake Fans",
            "40GBASE-T",
            Some("PoE++ Type 4 (Up to 90W)"),
            9,
            10,
        )
        .unwrap(),
        SubsystemSpec::new(
            "External Umbilical",
            "High-Speed Remote Uplink to Dedicated Workstation Hub",
            "40GBASE-T",
            None,
            11,
            12,
        )
        .unwrap(),
    ]
}

/// (Re)seed the canonical 12-port base chassis into the registry.
pub fn seed_base_registry() {
    let mut reg = registry().lock().unwrap();
    for spec in base_specs() {
        if !reg.iter().any(|e| e.name == spec.name) {
            reg.push(spec);
        }
    }
}

/// Register (or append) a subsystem spec. Port ranges must not overlap an
/// already-registered, differently-named spec.
pub fn register_subsystem(spec: SubsystemSpec) -> Result<(), String> {
    let mut reg = registry().lock().unwrap();
    for existing in reg.iter() {
        if existing.name == spec.name {
            continue;
        }
        let (a0, a1) = existing.port_range;
        let (b0, b1) = spec.port_range;
        if a0.max(b0) <= a1.min(b1) {
            return Err(format!(
                "Port range {:?} for '{}' overlaps '{}' range {:?}",
                spec.port_range, spec.name, existing.name, existing.port_range
            ));
        }
    }
    reg.push(spec);
    Ok(())
}

pub fn registered_subsystems() -> Vec<SubsystemSpec> {
    registry().lock().unwrap().clone()
}

pub fn clear_registry() {
    registry().lock().unwrap().clear();
}

#[derive(Debug, Clone, Serialize)]
pub struct PatchPort {
    pub port_id: u32,
    pub subsystem: String,
    pub interface_type: String,
    pub protocol: String,
    pub poe_class: Option<String>,
    pub link: Option<Cat8Link>,
}

impl std::fmt::Display for PatchPort {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let poe = self
            .poe_class
            .clone()
            .unwrap_or_else(|| "Unpowered".to_string());
        write!(
            f,
            "Port {:02} | {} | {} | {} | {}",
            self.port_id, self.subsystem, self.interface_type, self.protocol, poe
        )
    }
}

/// Programmatic builder for bespoke research rig layouts.
pub struct PatchPanelBuilder {
    specs: Vec<SubsystemSpec>,
}

impl PatchPanelBuilder {
    pub fn new() -> Self {
        Self { specs: Vec::new() }
    }

    pub fn add_subsystem(mut self, spec: SubsystemSpec) -> Self {
        self.specs.push(spec);
        self
    }

    pub fn build<F>(self, link_factory: F) -> PatchPanel
    where
        F: Fn(u32) -> Cat8Link,
    {
        PatchPanel::from_specs(&self.specs, link_factory)
    }
}

impl Default for PatchPanelBuilder {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone, Serialize, Default)]
pub struct PatchPanel {
    pub ports: Vec<PatchPort>,
}

impl PatchPanel {
    pub fn from_specs<F>(specs: &[SubsystemSpec], link_factory: F) -> Self
    where
        F: Fn(u32) -> Cat8Link,
    {
        let mut ports = Vec::new();
        for spec in specs {
            for pid in spec.port_ids() {
                ports.push(PatchPort {
                    port_id: pid,
                    subsystem: spec.name.clone(),
                    interface_type: spec.interface_type.clone(),
                    protocol: spec.protocol.clone(),
                    poe_class: spec.poe_class.clone(),
                    link: Some(link_factory(pid)),
                });
            }
        }
        ports.sort_by_key(|p| p.port_id);
        Self { ports }
    }

    /// Build the canonical 12-port base chassis (seeding if empty).
    pub fn default_layout<F>(link_factory: F) -> Self
    where
        F: Fn(u32) -> Cat8Link,
    {
        if registered_subsystems().is_empty() {
            seed_base_registry();
        }
        Self::from_specs(&registered_subsystems(), link_factory)
    }

    pub fn port(&self, port_id: u32) -> Option<&PatchPort> {
        self.ports.iter().find(|p| p.port_id == port_id)
    }

    pub fn subsystems(&self) -> Vec<String> {
        let mut seen = Vec::new();
        for p in &self.ports {
            if !seen.contains(&p.subsystem) {
                seen.push(p.subsystem.clone());
            }
        }
        seen
    }

    pub fn validate(&self) -> Vec<String> {
        let mut issues = Vec::new();
        for p in &self.ports {
            if let Some(link) = &p.link {
                issues.extend(link.validate());
            }
        }
        issues
    }

    pub fn report(&self) -> String {
        self.ports
            .iter()
            .map(|p| p.to_string())
            .collect::<Vec<_>>()
            .join("\n")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_layout_has_12_ports() {
        let panel = PatchPanel::default_layout(|pid| Cat8Link::new(format!("cat8-{pid:02}"), 4.5));
        assert_eq!(panel.ports.len(), 12);
        assert_eq!(
            panel.subsystems(),
            vec![
                "Head & Sensory Node".to_string(),
                "Upper Torso & Kinetic".to_string(),
                "Lower Base & Rig".to_string(),
                "Environmental Matrix".to_string(),
                "External Umbilical".to_string(),
            ]
        );
    }

    #[test]
    fn register_appends_node() {
        clear_registry();
        seed_base_registry();
        register_subsystem(
            SubsystemSpec::new(
                "R&D Lidar Array",
                "Solid-State Lidar + IMU",
                "40GBASE-T",
                Some("PoE++ Type 4 (Up to 90W)"),
                13,
                14,
            )
            .unwrap(),
        )
        .unwrap();
        let panel = PatchPanel::default_layout(|pid| Cat8Link::new(format!("c{pid}"), 3.0));
        assert!(panel.subsystems().contains(&"R&D Lidar Array".to_string()));
        assert_eq!(panel.ports.len(), 14);
        clear_registry();
    }

    #[test]
    fn register_rejects_overlap() {
        clear_registry();
        register_subsystem(
            SubsystemSpec::new(
                "A",
                "iface",
                "40GBASE-T",
                Some("PoE++ Type 4 (90W)"),
                13,
                16,
            )
            .unwrap(),
        )
        .unwrap();
        let res = register_subsystem(
            SubsystemSpec::new(
                "B",
                "iface",
                "40GBASE-T",
                Some("PoE++ Type 4 (90W)"),
                15,
                18,
            )
            .unwrap(),
        );
        assert!(res.is_err());
        clear_registry();
    }

    #[test]
    fn builder_bespoke_layout() {
        let panel = PatchPanelBuilder::new()
            .add_subsystem(
                SubsystemSpec::new("X", "i", "40GBASE-T", Some("PoE++ Type 4 (90W)"), 2, 2)
                    .unwrap(),
            )
            .add_subsystem(SubsystemSpec::new("Y", "i", "40GBASE-T", None, 5, 6).unwrap())
            .build(|pid| Cat8Link::new(format!("c{pid}"), 2.0));
        assert_eq!(
            panel.ports.iter().map(|p| p.port_id).collect::<Vec<_>>(),
            vec![2, 5, 6]
        );
        assert_eq!(panel.subsystems(), vec!["X".to_string(), "Y".to_string()]);
    }
}
